src/vmm/src/snapshot/mod.rs

```diff
@@ -66,46 +66,46 @@ pub enum SnapshotError {
 
 /// Firecracker snapshot header
 #[derive(Debug, Serialize, Deserialize)]
-struct SnapshotHdr {
+pub struct SnapshotHdr {
     /// magic value
     magic: u64,
     /// Snapshot data version
     version: Version,
 }
 
 impl SnapshotHdr {
-    fn new(version: Version) -> Self {
+    pub fn new(version: Version) -> Self {
         Self {
             magic: SNAPSHOT_MAGIC_ID,
             version,
         }
     }
-}
 
-/// Firecracker snapshot type
-///
-/// A type used to store and load Firecracker snapshots of a particular version
-#[derive(Debug)]
-pub struct Snapshot {
-    // The snapshot version we can handle
-    version: Version,
-}
+    pub fn load<R: Read>(reader: &mut R) -> Result<Self, SnapshotError> {
+        let hdr: SnapshotHdr = bincode::DefaultOptions::new()
+            .with_limit(VM_STATE_DESERIALIZE_LIMIT)
+            .with_fixint_encoding()
+            .allow_trailing_bytes() // need this because we deserialize header and snapshot from the same file, so after
+            // reading the header, there will be trailing bytes.
+            .deserialize_from(reader)
+            .map_err(|err| SnapshotError::Serde(err.to_string()))?;
 
-impl Snapshot {
-    /// Creates a new instance which can only be used to save a new snapshot.
-    pub fn new(version: Version) -> Snapshot {
-        Snapshot { version }
+        Ok(hdr)
     }
 
-    /// Fetches snapshot data version.
-    pub fn get_format_version<T>(reader: &mut T) -> Result<Version, SnapshotError>
-    where
-        T: Read + Debug,
-    {
-        let hdr: SnapshotHdr = Self::deserialize(reader)?;
-        Ok(hdr.version)
+    pub fn store<W: Write>(&self, writer: &mut W) -> Result<(), SnapshotError> {
+        bincode::serialize_into(writer, self).map_err(|err| SnapshotError::Serde(err.to_string()))
     }
+}
 
+#[derive(Debug, Serialize, Deserialize)]
+pub struct Snapshot<Data> {
+    // The snapshot version we can handle
+    header: SnapshotHdr,
+    pub data: Data,
+}
+
+impl<Data: for<'a> Deserialize<'a>> Snapshot<Data> {
     /// Helper function to deserialize an object from a reader
     pub fn deserialize<T, O>(reader: &mut T) -> Result<O, SnapshotError>
```

Comment:
This function should probably be free-standing, and not `pub`. Then you can reuse it inside `SnapshotHdr::load` as well

Reply:
By free standing I'm assuming you mean not have the deserialize and serialize functions as implementations of Snapshot so I moved them outside.